README.md
# Name
    libvirt-prometheus-exporter - a prometheus exporter for libvirt

# Synopsis
    libvirt-prometheus-exporter {http-port} {libvirt-ipc-url}

# Description
    
    http-port
        the http-port to listen on.
    libvirt-ipc-url
        e.g qemu:///system

# Environment for systemd
    /etc/default/libvirt-prometheus-exporter
    

# Build and test
- libvirt (> 8.0)
- debhelper (>9.0)
- g++ (>10)
- Make
- devscripts

## build
    bash devops/build.sh

### Changelog
    remember to update the changelog in debian/changelog

debian/changelog
libvirt-prometheus-exporter (0.2.12) jammy; urgency=low

  * BUGFIX: Added libvirt_up and libvirt_requests

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 7 Nov 2023 21:54:37 +0000

libvirt-prometheus-exporter (0.2.11) jammy; urgency=low

  * BUGFIX: Added tenancy uuid from metadata and libvirt_number_up

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 7 Nov 2023 20:48:37 +0000

libvirt-prometheus-exporter (0.2.10) jammy; urgency=low

  * BUGFIX: Rename labels to comply with namespace-relevance.

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 18 Sep 2023 20:33:37 +0000

libvirt-prometheus-exporter (0.2.9) jammy; urgency=low

  * CHORE: Rename labels to comply with namespace-relevance.

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 18 Sep 2023 17:32:37 +0000

libvirt-prometheus-exporter (0.2.8) jammy; urgency=low

  * BUGFIX: Wrong ownership of files.

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 17 Sep 2023 12:56:37 +0000

libvirt-prometheus-exporter (0.2.7) jammy; urgency=low

  * CHORE: Cleanup network-metrics.

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 17 Sep 2023 12:44:37 +0000

libvirt-prometheus-exporter (0.2.6) jammy; urgency=low

  * FEATURE: Added network metrics

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 16 Sep 2023 20:48:37 +0000

libvirt-prometheus-exporter (0.2.5) jammy; urgency=low

  * BUGFIX: sanitize build, change signing key, moved into /usr/sbin

 -- Newsworthy39 <newsworthy39@github.com>  Thu, 16 Sep 2023 15:32:37 +0000

libvirt-prometheus-exporter (0.2.4) jammy; urgency=low

  * BUGFIX: Deps

 -- Michael Jensen <michaeljensendk@hotmail.com>  Thu, 15 Sep 2023 20:30:37 +0000

libvirt-prometheus-exporter (0.2.3) jammy; urgency=low

  * BUGFIX: Made exporter exit, when the endpoint is not available.

 -- Michael Jensen <michaeljensendk@hotmail.com>  Thu, 15 Sep 2023 20:12:37 +0000

libvirt-prometheus-exporter (0.2.2) jammy; urgency=low

  * Added missing build-dependecy for libvirt-dev

 -- Michael Jensen <michaeljensendk@hotmail.com>  Thu, 14 Sep 2023 21:35:37 +0000

libvirt-prometheus-exporter (0.2.1) jammy; urgency=low

  * Updated documentation for better viewing and
    making it more readable online in markdown.

 -- Michael Jensen <michaeljensendk@hotmail.com>  Thu, 14 Sep 2023 20:20:37 +0000

libvirt-prometheus-exporter (0.2.0) jammy; urgency=low

  * Changed binary package name

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 20:20:37 +0000

libvirtprometheusexporter (0.1.2) jammy; urgency=low

  * Bugfix: Environmentfile was written errorneus

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 20:12:37 +0000

libvirtprometheusexporter (0.1.1) jammy; urgency=low

  * Added systemd integration

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 20:03:37 +0000

libvirtprometheusexporter (0.1.0) jammy; urgency=low

  * Added man-pages and other documentation to binary-package.

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 19:54:37 +0000

libvirtprometheusexporter (0.0.2) jammy; urgency=low

  * Improved build-environment for packing.

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 18:54:37 +0000


libvirtprometheusexporter (0.0.1) jammy; urgency=low

  * Initial release.

 -- Newsworthy39 <newsworthy39@launchpad.net>  Thu, 14 Sep 2023 14:25:37 +0000


test.php
<?php
     $conn = libvirt_connect('null', false);
     $doms = libvirt_list_domains($conn);
     print_r($doms);


     libvirt_image_create($conn, 'test.img',8192, "qcow2");
?>


test/client.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <nanomsg/nn.h>
#include <nanomsg/pubsub.h>

#define SERVER "server"
#define CLIENT "client"

int
client(const char *url, const char *name)
{
        int sock;

        if ((sock = nn_socket(AF_SP, NN_SUB)) < 0) {
                 fprintf(stderr,"nn_socket");
        }

        // subscribe to everything ("" means all topics)
        if (nn_setsockopt(sock, NN_SUB, NN_SUB_SUBSCRIBE, "", 0) < 0) {
                 fprintf(stderr,"nn_setsockopt");
        }
        if (nn_connect(sock, url) < 0) {
                fprintf(stderr, "nn_connet");
        }
        for (;;) {
                char *buf = NULL;
                int bytes = nn_recv(sock, &buf, NN_MSG, 0);
                if (bytes < 0) {
                         fprintf(stderr,"nn_recv");
                }
                printf("CLIENT (%s): RECEIVED %s\n", name, buf); 
                nn_freemsg(buf);
        }
}

int
main(const int argc, const char **argv)
{
          if ((argc >= 3) && (strcmp(CLIENT, argv[1]) == 0))
                return (client (argv[2], argv[3]));

        fprintf(stderr, "Usage: pubsub %s|%s <URL> <ARG> ...\n",
            SERVER, CLIENT);
        return 1;
}

test/info1.cpp
/**
 * section: Information
 * synopsis: Extract information about Xen domain 0
 * purpose: Demonstrate the basic use of the library to connect to the
 *          hypervisor and extract domain information.
 * usage: info1
 * test: info1
 * copy: see Copyright for the status of this software.
 */

#include <iostream>
#include <algorithm>
#include <thread>
#include <stdio.h>
#include <chrono>
#include <libvirt/libvirt.h>
#include <nanomsg/pubsub.h>
#include <nanomsg/nn.h>
#include <string.h>

static const std::vector<std::string>
getDomains(const virConnectPtr &conn)
{
    virDomainPtr *domains;
    unsigned int flags = VIR_CONNECT_LIST_DOMAINS_RUNNING | VIR_CONNECT_LIST_DOMAINS_PAUSED;
    size_t i;
    int ret;
    std::vector<std::string> str_domains;

    ret = virConnectListAllDomains(conn, &domains, flags);
    if (ret < 0)
        fprintf(stderr, "No domains found.");

    for (i = 0; i < ret; i++)
    {
        /* Get the information, guaranted to have a name */

        const char *name = virDomainGetName(domains[i]);
        str_domains.push_back(name);

        // here or in a separate loop if needed
        virDomainFree(domains[i]);
    }

    if (domains != NULL)
        free(domains);

    return str_domains;
}

template <typename T>
std::vector<T> vectorDifference(const std::vector<T> &v1, const std::vector<T> &v2)
{
    // make the result initially to v1, but only if its not empty
    std::vector<T> result = v1;

    // remove the elements in v2 from result
    for (const T &element : v2)
    {
        const auto it = std::find_if(result.begin(), result.end(), [element](const T &node)
                                     { return node == element; });

        if (it != result.end())
        {
            result.erase(it);
        }
    }
    return result;
}

/**
 * @brief This server function, serves as a guard, initializing and proxying the nn_send-function.
 *        This function blocks indefinitely.
 * 
 * @param url the url to bind to
 * @param fnc lambda, to send back to local lambda
 * @return int 0
 */
int server(const char *url, std::function<void(std::function<void(const char *msg, int length)>)> fnc)
{
    // Allways populate at the beginning, to avoid unnecessary
    // state change messages.
    int sock = 0;
    if ((sock = nn_socket(AF_SP, NN_PUB)) < 0)
    {
        fprintf(stderr, "nn_socket");
    }
    if (nn_bind(sock, url) < 0)
    {
        fprintf(stderr, "nn_bind");
    }

    for (;;)
    {
        // send messages, if any. We use a proxy to avoid refencing local/stack-variables and 
        // to avoid the heap.
        auto snd = [&sock](const char *msg, int length)
        {
            int bytes = nn_send(sock, msg, length, 0);
            if (bytes < 0)
            {
                fprintf(stderr, "nn_send");
            }
        };

        fnc(snd); // dumpster fires are fun.

        using namespace std::chrono_literals;
        std::this_thread::sleep_for(1000ms);
    }
}

int main(int argc, char **argv)
{
    if (argc != 3)
    {
        fprintf(stderr, "syntax: %s: ipc-uri libvirt-uri", argv[0]);
        return 1;
    }

    virConnectPtr conn = virConnectOpenReadOnly(argv[2]);
    if (conn == NULL)
    {
        fprintf(stderr, "Failed to connect to hypervisor\n");
    }

    // We use lambdas to fix scoping issues and avoid having to work with far-memory.
    // We send a send-proxy alongside, to avoid copying/referencing memory outside
    // of the segment.
    std::vector<std::string> str_domains = getDomains(conn);
    server(argv[1], [&conn, &str_domains](std::function<void(const char *msg, int length)> send) -> void
           {
               std::vector<std::string> tmp = getDomains(conn);
               std::vector<std::string> diff;
               int on = 1;

               if (tmp.size() <= str_domains.size())
               {
                   // machines are leaving
                   diff = vectorDifference(str_domains, tmp);
                   on = 0;
               }
               else
               {
                   diff = vectorDifference(tmp, str_domains);
               }

               for (auto i : diff)
               {
                    char buffer[4096] = { 0 };
                    size_t sz_d = snprintf(buffer, 4096, "{ \"%s\": \"state\": %d }", i.c_str(), on);
                    send(buffer,sz_d );                   
               }

               str_domains = tmp;
           });

    if (conn != NULL)
        virConnectClose(conn);

    return 0;
}


info1.cpp
/**
 * section: Information
 * synopsis: Extract information about Xen domain 0
 * purpose: Demonstrate the basic use of the library to connect to the
 *          hypervisor and extract domain information.
 * usage: info1
 * test: info1
 * copy: see Copyright for the status of this software.
 */

#include <iostream>
#include <algorithm>
#include <thread>
#include <stdio.h>
#include <chrono>
#include <libvirt/libvirt.h>
#include <nanomsg/pubsub.h>
#include <nanomsg/nn.h>
#include <string.h>

static const std::vector<std::string>
getDomains(const virConnectPtr &conn)
{
    virDomainPtr *domains;
    unsigned int flags = VIR_CONNECT_LIST_DOMAINS_RUNNING | VIR_CONNECT_LIST_DOMAINS_PAUSED;
    size_t i;
    int ret;
    std::vector<std::string> str_domains;

    ret = virConnectListAllDomains(conn, &domains, flags);
    if (ret < 0)
        fprintf(stderr, "No domains found.");

    for (i = 0; i < ret; i++)
    {
        /* Get the information, guaranted to have a name */

        const char *name = virDomainGetName(domains[i]);
        str_domains.push_back(name);

        // here or in a separate loop if needed
        virDomainFree(domains[i]);
    }

    if (domains != NULL)
        free(domains);

    return str_domains;
}

template <typename T>
std::vector<T> vectorDifference(const std::vector<T> &v1, const std::vector<T> &v2)
{
    // make the result initially to v1, but only if its not empty
    std::vector<T> result = v1;

    // remove the elements in v2 from result
    for (const T &element : v2)
    {
        const auto it = std::find_if(result.begin(), result.end(), [element](const T &node)
                                     { return node == element; });

        if (it != result.end())
        {
            result.erase(it);
        }
    }
    return result;
}

int server(const char *url, std::function<void(std::function<void(const char *msg, int length)>)> fnc)
{
    // Allways populate at the beginning, to avoid unnecessary
    // state change messages.
    int sock = 0;
    if ((sock = nn_socket(AF_SP, NN_PUB)) < 0)
    {
        fprintf(stderr, "nn_socket");
    }
    if (nn_bind(sock, url) < 0)
    {
        fprintf(stderr, "nn_bind");
    }

    for (;;)
    {
        // send messages, if any.
        auto snd = [&sock](const char *msg, int length)
        {
            int bytes = nn_send(sock, msg, length, 0);
            if (bytes < 0)
            {
                fprintf(stderr, "nn_send");
            }
        };

        fnc(snd);

        // wait abit
        using namespace std::chrono_literals;
        std::this_thread::sleep_for(1000ms);
    }
}

int main(int argc, char **argv)
{
    if (argc != 3)
    {
        fprintf(stderr, "syntax: %s: ipc-uri libvirt-uri", argv[0]);
        return 1;
    }

    virConnectPtr conn = virConnectOpenReadOnly(argv[2]);
    if (conn == NULL)
    {
        fprintf(stderr, "Failed to connect to hypervisor\n");
    }

    // We use lambdas to fix scoping issues and avoid having to work with far-memory.
    // We send a send-proxy alongside, to avoid copying/referencing memory outside
    // of the segment.
    std::vector<std::string> str_domains = getDomains(conn);
    server(argv[1], [&conn, &str_domains](std::function<void(const char *msg, int length)> send) -> void
           {
               std::vector<std::string> tmp = getDomains(conn);
               std::vector<std::string> diff;
               int on = 1;

               if (tmp.size() <= str_domains.size())
               {
                   // machines are leaving
                   diff = vectorDifference(str_domains, tmp);
                   on = 0;
               }
               else
               {
                   diff = vectorDifference(tmp, str_domains);
               }

               for (auto i : diff)
               {
                    char buffer[4096] = { 0 };
                    size_t sz_d = snprintf(buffer, 4096, "{ \"%s\": \"state\": %d }", i.c_str(), on);
                    send(buffer,sz_d );                   
               }

               str_domains = tmp;
           });

    if (conn != NULL)
        virConnectClose(conn);

    return 0;
}


include/format.h
#include <memory>
#include <string>
#include <stdexcept>

namespace custom
{
    template <typename... Args>
    std::string format(const std::string &format, Args... args)
    {
        int size_s = std::snprintf(nullptr, 0, format.c_str(), args...) + 1; // Extra space for '\0'
        if (size_s <= 0)
        {
            throw std::runtime_error("Error during formatting.");
        }
        auto size = static_cast<size_t>(size_s);
        std::unique_ptr<char[]> buf(new char[size]);
        std::snprintf(buf.get(), size, format.c_str(), args...);
        return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
    }

    /**
     * @brief Helper to generate response
     *
     * @param reply
     * @param response
     * @param version
     * @return std::string
     */
    std::string generate_prometheus(std::string reply, int response = 200, std::string version = "HTTP/1.0")
    {
        std::string type = "text/plain; version=0.0.4";
        std::string status = "200 OK";
        return custom::format("%s %s\r\nContent-Length: %d\r\nContent-Type: %s\r\n\r\n%s", version.c_str(),
                              status.c_str(), reply.length(), type.c_str(), reply.c_str());
    }

    std::vector<std::string> split(const std::string &s, std::string delimiter = ".")
    {
        // for string delimiter
        size_t pos_start = 0, pos_end, delim_len = delimiter.length();
        std::string token;
        std::vector<std::string> res;

        while ((pos_end = s.find(delimiter, pos_start)) != std::string::npos)
        {
            token = s.substr(pos_start, pos_end - pos_start);
            pos_start = pos_end + delim_len;
            res.push_back(token);
        }

        res.push_back(s.substr(pos_start));

        return res;
    }
}

include/format.hpp
#ifndef __FORMAT_HPP__
#define __FORMAT_HPP__

#include <memory>
#include <string>
#include <stdexcept>
#include <vector>
#include <libvirt/libvirt.h>

namespace custom
{

    template <typename... Args>
    inline std::string format(const std::string &format, Args... args)
    {
        int size_s = std::snprintf(nullptr, 0, format.c_str(), args...) + 1; // Extra space for '\0'
        if (size_s <= 0)
        {
            throw std::runtime_error("Error during formatting.");
        }
        auto size = static_cast<size_t>(size_s);
        std::unique_ptr<char[]> buf(new char[size]);
        std::snprintf(buf.get(), size, format.c_str(), args...);
        return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
    }

    /**
     * @brief Helper to generate response
     *
     * @param reply
     * @param response
     * @param version
     * @return std::string
     */

    inline std::string generate_prometheus(std::string reply, int response = 200, std::string version = "HTTP/1.0")
    {
        std::string type = "text/plain; version=0.0.4";
        std::string status = "200 OK";
        return custom::format("%s %s\r\nContent-Length: %d\r\nContent-Type: %s\r\n\r\n%s", version.c_str(),
                              status.c_str(), reply.length(), type.c_str(), reply.c_str());
    }

    inline std::vector<std::string> split(const std::string &s, std::string delimiter = ".")
    {
        // for string delimiter
        size_t pos_start = 0, pos_end, delim_len = delimiter.length();
        std::string token;
        std::vector<std::string> res;

        while ((pos_end = s.find(delimiter, pos_start)) != std::string::npos)
        {
            token = s.substr(pos_start, pos_end - pos_start);
            pos_start = pos_end + delim_len;
            res.push_back(token);
        }

        res.push_back(s.substr(pos_start));

        return res;
    }

    inline int virDomainGetTenant(virDomainPtr domain, char *buf)
    {
        // Get metadata,
        const char *domain_meta_xml = virDomainGetMetadata(domain,
                                                           virDomainMetadataType::VIR_DOMAIN_METADATA_ELEMENT,
                                                           "http://portfolio.org/virtualization/instance",
                                                           virDomainModificationImpact::VIR_DOMAIN_AFFECT_CURRENT);
        if (domain_meta_xml != NULL) {
            strncpy(buf, &domain_meta_xml[32], 36); // TODO: horrible.
            return 1;
        }

        return 0;
    }

}

#endif

include/serializers.hpp
#ifndef __VCPU_SERIALIZER_HPP__
#define __VCPU_SERIALIZER_HPP__

#include <iostream>
#include <string>
#include <string.h>
#include <format.hpp>
#include <vector>
#include <libvirt/libvirt.h>

namespace serializer
{
    inline void vcpu_metrics(std::string &out, size_t rc, virDomainStatsRecordPtr *stats)
    {
        for (size_t j = 0; j < rc; j++)
        {
            virDomainStatsRecordPtr record = stats[j];

            for (int k = 0; k < record->nparams; k++)
            {

                char tenantuuid[4096] = {0};
                custom::virDomainGetTenant(record->dom, &tenantuuid[0]);

                char uuid[4096] = {0};
                virDomainGetUUIDString(record->dom, uuid);

                // <instance>
                // <tenant>
                // <uuid>43dc0cf8-809b-4adb-9bea-a9abb5f3d90e</uuid>
                // </tenant>
                // </instance>

                // input is vcpu, id, param
                std::vector<std::string> fields = custom::split(record->params[k].field);
                if (fields.size() == 3)
                {
                    out.append(custom::format("# TYPE libvirt_%s_%s counter\n", fields[0].c_str(), fields[2].c_str()));
                    out.append(custom::format("libvirt_%s_%s{domain=\"%s\", vcpu=\"%s\" uuid=\"%s\" tenant=\"%s\"} %llu\n",
                                              fields[0].c_str(),
                                              fields[2].c_str(),
                                              virDomainGetName(record->dom),
                                              fields[1].c_str(),
                                              uuid,
                                              tenantuuid,
                                              record->params[k].value.ul));
                }
            }
        }
    }

    inline void network_metrics(std::string &out, size_t rc, virDomainStatsRecordPtr *stats)
    {
        for (size_t j = 0; j < rc; j++)
        {
            virDomainStatsRecordPtr record = stats[j];
            std::string netname;

            char tenantuuid[4096] = {0};
            custom::virDomainGetTenant(record->dom, &tenantuuid[0]);

            char uuid[4096] = {0};
            virDomainGetUUIDString(record->dom, uuid);

            for (int k = 0; k < record->nparams; k++)
            {

                // input is vcpu, id, param
                std::vector<std::string> fields = custom::split(record->params[k].field);
                if (fields.size() == 3)
                {
                    netname = record->params[k].value.s;
                }

                if (fields.size() == 4)
                {
                    out.append(custom::format("# TYPE libvirt_%s_%s_%s counter\n", fields[0].c_str(), fields[3].c_str(), fields[2].c_str()));
                    out.append(custom::format("libvirt_%s_%s_%s{domain=\"%s\" interfaceid=\"%s\", name=\"%s\" uuid=\"%s\" tenant=\"%s\"} %llu\n",
                                              fields[0].c_str(), fields[3].c_str(), fields[2].c_str(), // => net_bytes_rx
                                              virDomainGetName(record->dom), fields[1].c_str(), netname.c_str(),
                                              uuid,
                                              tenantuuid,
                                              record->params[k].value.ul));
                }
            }
        }
    }

    inline void block_metrics(std::string &out, size_t rc, virDomainStatsRecordPtr *stats)
    {
        for (size_t j = 0; j < rc; j++)
        {
            virDomainStatsRecordPtr record = stats[j];
            char tenantuuid[4096] = {0};
            custom::virDomainGetTenant(record->dom, &tenantuuid[0]);

            char uuid[4096] = {0};
            virDomainGetUUIDString(record->dom, uuid);

            for (int k = 0; k < record->nparams; k++)
            {
                // input is vcpu, id, param
                std::vector<std::string> fields = custom::split(record->params[k].field);
                if (fields.size() == 4)
                {
                    out.append(custom::format("# TYPE libvirt_%s_%s_%s counter\n", fields[0].c_str(), fields[3].c_str(), fields[2].c_str()));
                    out.append(custom::format("libvirt_%s_%s_%s{domain=\"%s\" blockid=\"%s\", uuid=\"%s\" tenant=\"%s\"} %llu\n",
                                              fields[0].c_str(), fields[3].c_str(), fields[2].c_str(), // => net_bytes_rx
                                              virDomainGetName(record->dom), fields[1].c_str(), 
                                              uuid,
                                              tenantuuid,
                                              record->params[k].value.ul));
                }
            }
        }
    }

}

#endif

src/prometheus.cpp
/**
 * @file prometheus.cpp
 * @author Newsworthy39 <newsworthy39@github.com>
 * @brief Exports prometheus metrics from libvirt vms.
 * @version 0.2.5
 * @date Thu, 16 Sep 2023 15:32:37 +0000
 *
 * @copyright Copyright (c) 2023 Newsworthy39 <newsworthy39@github.com>
 *
 */

#include <iostream>
#include <algorithm>
#include <string>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/epoll.h>
#include <unistd.h>
#include <format.hpp>
#include <serializers.hpp>
#include <libvirt/libvirt.h>

#define MAX_EVENTS 10
#define BACKLOG 10

/**
 * @brief setnonblocking
 * sets a socket into nonblocking.
 *
 * @param sock
 * @return int
 */
int setnonblocking(int sock)
{
    int result;
    int flags;

    flags = ::fcntl(sock, F_GETFL, 0);

    if (flags == -1)
    {
        return -1; // error
    }

    flags |= O_NONBLOCK;

    result = fcntl(sock, F_SETFL, flags);
    return result;
}

/**
 * @brief setflags,
 * Sets different flags on a socket.
 *
 * @param sock
 */
void setflags(int sock)
{
    int optval = 1;
    setsockopt(sock, SOL_SOCKET, SO_REUSEPORT, &optval, sizeof(optval));
    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
}

/**
 * @brief stream_server
 * This is an event-based, stack-based stream-server. It basically,
 * just spews any data onto a callable, with a FD-on it.
 * @param port the port to listen on
 * @param fnc the callback
 * @return int
 */

int stream_server(int port, std::function<void(int)> fnc)
{
    struct epoll_event ev, events[MAX_EVENTS];
    int sockfd, epollfd, n;
    struct sockaddr_in serv_addr; /* my address information */

    /* Code to set up listening socket, 'listen_sock',
       (socket(), bind(), listen()) omitted. */
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1)
    {
        perror("socket");
        exit(1);
    }

    serv_addr.sin_family = AF_INET;         /* host byte order */
    serv_addr.sin_port = htons(port);       /* short, network byte order */
    serv_addr.sin_addr.s_addr = INADDR_ANY; /* auto-fill with my IP */
    bzero(&(serv_addr.sin_zero), 8);        /* zero the rest of the struct */

    if (bind(sockfd, (struct sockaddr *)&serv_addr, sizeof(struct sockaddr)) == -1)
    {
        perror("bind");
        exit(1);
    }

    setflags(sockfd);

    if (listen(sockfd, BACKLOG) == -1)
    {
        perror("listen");
        exit(1);
    }

    /* epoll*/
    epollfd = epoll_create1(0);
    if (epollfd == -1)
    {
        perror("epoll_create1");
        exit(EXIT_FAILURE);
    }

    ev.events = EPOLLIN;
    ev.data.fd = sockfd;
    if (epoll_ctl(epollfd, EPOLL_CTL_ADD, sockfd, &ev) == -1)
    {
        perror("epoll_ctl: listen_sock");
        exit(EXIT_FAILURE);
    }

    while (true)
    {
        int nfds = epoll_wait(epollfd, events, MAX_EVENTS, -1);
        if (nfds == -1)
        {
            perror("epoll_wait");
            exit(EXIT_FAILURE);
        }

        for (n = 0; n < nfds; ++n)
        {
            if (events[n].data.fd == sockfd)
            {
                socklen_t addrlen;
                struct sockaddr_in client_addr; /* my address information */

                int conn_sock = accept(events[n].data.fd,
                                       (struct sockaddr *)&client_addr, &addrlen);
                if (conn_sock == -1)
                {
                    fprintf(stderr, "accept: %s\n", strerror(errno));
                    exit(EXIT_FAILURE);
                }

                setnonblocking(conn_sock);
                ev.events = EPOLLIN | EPOLLET;
                ev.data.fd = conn_sock;
                if (epoll_ctl(epollfd, EPOLL_CTL_ADD, conn_sock,
                              &ev) == -1)
                {
                    perror("epoll_ctl: conn_sock");
                    exit(EXIT_FAILURE);
                }
            }
            else
            {
                // We have data, to read -
                // send messages, if any. We use a proxy to avoid refencing local/stack-variables and
                // to avoid the heap.
                fnc(events[n].data.fd);
            }
        }
    }
}

/**
 * @brief Main entry point
 *
 * @param argc Number of arguments
 * @param argv the actual arguments
 * @return int
 */
int main(int argc, char **argv)
{
    long long requests = 0;

    if (argc != 3)
    {
        fprintf(stderr, "syntax: %s: http-port libvirt-uri\n", argv[0]);
        return 1;
    }

    int port = std::stoi(std::string(argv[1]));

    printf("using port: %d\n", port);
    printf("using system: %s\n", argv[2]);

    virConnectPtr conn = virConnectOpenReadOnly(argv[2]);
    if (conn == NULL)
    {
        fprintf(stderr, "Failed to connect to %s\n", argv[2]);
        return (EXIT_FAILURE);
    }

    // Setup a stream_server, and use the lambda below
    // for data-processing.
    stream_server(port, [&conn, &requests](int fd) -> void {
        // read from socket, consume data.
        char buffer[4096] = {0};
        ssize_t bytes = recv(fd, buffer, 4096, 0);

        // Deal with different reading-scenarios.
        if (bytes > 0) {

            // Generate output
            std::string body;
            virDomainPtr *doms = { 0 };
            body.append("# prometheus data\n");

            // List, domains and take stats, but only if something is there.
            unsigned int flags = VIR_CONNECT_LIST_DOMAINS_RUNNING;
            size_t ret = virConnectListAllDomains(conn, &doms, flags);
            if (ret <= 0) {
                std::string output = custom::generate_prometheus(body);
                send(fd, output.c_str(), output.length(), 0);
                return;
            }

            // cpu-stats.
            virDomainStatsRecordPtr *statscpu;
            size_t rc = virDomainListGetStats(doms,
                                            virDomainStatsTypes::VIR_DOMAIN_STATS_VCPU,
                                            &statscpu, VIR_CONNECT_GET_ALL_DOMAINS_STATS_NOWAIT);
                                            
            if (rc > 0)
            {
                serializer::vcpu_metrics(body, rc, statscpu);
            }
            // Free structures
            virDomainStatsRecordListFree(statscpu);             

            // Network stats
            virDomainStatsRecordPtr *statsnet;
            size_t nrc = virDomainListGetStats(doms,
                                            virDomainStatsTypes::VIR_DOMAIN_STATS_INTERFACE,
                                            &statsnet, VIR_CONNECT_GET_ALL_DOMAINS_STATS_NOWAIT);
            if (nrc > 0)
            {
                serializer::network_metrics(body, nrc, statsnet);                
            }
            // Free structures
            virDomainStatsRecordListFree(statsnet);

             // block stats
            virDomainStatsRecordPtr *statsblocks;
            size_t nrb = virDomainListGetStats(doms,
                                            virDomainStatsTypes::VIR_DOMAIN_STATS_BLOCK,
                                            &statsblocks, VIR_CONNECT_GET_ALL_DOMAINS_STATS_NOWAIT);
            if (nrb > 0)
            {
                serializer::block_metrics(body, nrc, statsblocks);                
            }
            // Free structures
            virDomainStatsRecordListFree(statsblocks);

            // Make "up
            for (size_t j = 0; j < ret; j++) {
            
                char tenantuuid[4096] = {0};
                custom::virDomainGetTenant(*(&doms[j]), &tenantuuid[0]);

                char uuid[4096] = {0};
                virDomainGetUUIDString(*(&doms[j]), uuid);

                body.append(custom::format("# TYPE libvirt_%s gauge\n", "up"));
                body.append(custom::format("libvirt_up{domain=\"%s\" uuid=\"%s\" tenant=\"%s\"} 1\n",
                                              virDomainGetName(*(&doms[j])), 
                                              uuid,
                                              tenantuuid));
            }

            // Stats about the exporter
            body.append(custom::format("# TYPE libvirt_%s counter\n", "requests"));
            body.append(custom::format("libvirt_requests %ld\n", ++requests));

            // allways free here.            
            free(doms);

            std::string output = custom::generate_prometheus(body);
            send(fd, output.c_str(), output.length(), 0);
        }

        if (bytes <= 0)
            close(fd); 
    });

    if (conn != NULL)
        virConnectClose(conn);

    return 0;
}